accepted/2024/trusted-publishers-oidc-for-nuget-push.md

```diff
@@ -0,0 +1,523 @@
+
+#### What NuGet API operations can be used inside the GitHub Actions workflow?
+
+To scope the ways in which a GitHub trust policy can be used, similar scoping rules to the current API key flow can be
+set on the GitHub Actions trust policy:
+
+![Scope form](meta/trusted-publishers-oidc-for-nuget-push/scope-form.png)
+
+Just like the existing API scoping feature, there are 3 dimensions which a trust policy can be scoped. For more
+information about scopes on API keys, see [Scoped API
+keys](https://learn.microsoft.com/en-us/nuget/nuget-org/scoped-api-keys). In summary:
+
+- **Package owner**: for users that are members of one or more organizations, it's important to know which package owner
+  this trust policy acts on behalf of. For example, in my screenshot above the package owner is set to `joelverhagen`,
+  so packages owned only by my `snoozecorp` organization will not accept package uploads via this trust policy. This is
+  also important to define who is the initial owner of a package that is newly created. In this example, if I push a new
+  `Knapcode.MyBestPackage` package ID, the new package ID will be owned by `joelverhagen`, not `snoozecorp` or any other
+  organization that I am a member of based on this package owner definition. This resolves the ["Pending" Trusted
+  Publishers](https://repos.openssf.org/trusted-publishers-for-all-package-repositories#pending-trusted-publishers)
+  problem.
+
+- **Scopes**: the scope radio buttons and checkboxes define what kind of operations the trust policy can be used for.
+  For example, can the trust policy be used to unlist packages in addition to publishing new packages? Can only versions
+  on existing package IDs be published? Or new package IDs be created as well?
+
+- **Glob pattern**: this defines which package IDs the trust policy can apply to, based on a wildcard pattern (called
+  "glob" in this context for historical reasons). If you want the trust policy to apply to all package IDs you own, you
+  would specify `*`. Note that a broad wildcard like `*` does not supersede access controls on package IDs you do not
+  own. It just defines which if your packages the trust policy can work for.
+
+The screenshot above allows only new package versions to be pushed for packages owned by `joelverhagen`, having IDs
+starting in `Knapcode.`. New package IDs cannot be created in this case nor can packages be unlisted.
+
+Unlike API key definitions, there is no "Expires In" field, because a trust policy never expires.
+
+Multiple trust policies can be defined. If any of the trust policies match the incoming package and GitHub Action token,
+the operation will be allowed. If the package ID is new and there is ambiguity on which package owner should be assigned
+to the package (due to multiple matching trust policies), the most recently created trust policy will be used.
+
+### Step 2: add a GitHub Actions workflow with NuGet.org Trusted Publisher authentication
```

Comment:
I assume we will want to have a NuGet GH action template on the marketplace to simplify this beyond docs?

Reply:
Are you referring to a workflow template like this?
https://docs.github.com/en/actions/using-workflows/creating-starter-workflows-for-your-organization
Or a GitHub Action to do pack + auth + push all in one?

Reply:
More so on the public marketplace:

https://github.com/rubygems/release-gem?tab=readme-ov-file#requirements -> https://github.com/marketplace/actions/release-gem

Reply:
The `nuget/login` action in the design would be on the public marketplace. But it would only do the OIDC -> short-lived API key trade, and not the push step. ... but maybe it could? It would need a parameter to specify which .nupkgs to push. 

I think it's better to keep the public marketplace action to be minimal, and re-usable for multiple use cases. If the action does `nuget push` internally, then we may end up needing all of the CLI parameters that go into `nuget push`, and add them as push gets more complex.

I would prefer `nuget/login` public marketplace action and not `nuget/release` for now. This also makes using the short-lived API key for unlist/relist and even deprecate API easier since all `nuget/login` does is produce an API key. It's not hardcode to always push.